### ISSUE TITLE
Useful function for strings

### ISSUE THREAD

**Issue by Antidote** [LordNed/WindEditor#20]
I discovered how Nintendo hashes file/directory names in RARC files on kuribo, and decided to apply it to the joint names of BMD/BDL files, and it worked perfectly.

Here is the code I found (converted to C#):

``` csharp
ushort calculateHash(string target)
{
    ushort hash = 0;
    foreach (char c in target)
    {
        hash *= 3;
        hash += (ushort)c;
    }

    return hash;
}
```

Rather than make a pull request for it I decided to just make an issue and give you the function. As I'm not familiar with your codebase, nor am I familiar enough with C# anymore to attempt to become familiar with it. I've already updated the relevant wiki article for it.

**Comment by LordNed** [comment]
Hey,

Thanks for the PR/Issue! :D

Do you happen to have any idea why Nintendo has string hashes in either
RARC or BMD/BDL? I've been notified by other devs that the tools used to
re-create arc files never did the hashes and the engine didn't care, and it
doesn't seem to care about the hashes in the BMD/BDL files either, so I'm
curious as to why they bothered putting them in in the first place.

(I'll leave the issue open for now until I get time to get this integrated
somewhere in the codebase, thanks for the contribution!)

Thanks,
LordNed

**Comment by Antidote** [comment]
It was probably a way to expedite searches: Hash the string, and check for a match that way. And if it doesn't have one do it the slower route (string comparison).

**Comment by NWPlayer123** [comment]
Yeah. I know for sure in data formats like BCSV (Binary CSV, basically a table), the hash is used by the game. An example of this is Super Mario Galaxy 2 using ScenarioData's last few fields (with hashes) to load different zones, I can hash a new zone name, add in a new field, and the game will then find the folder with that hash and read/use it. As for RARC, IDK if the engine particularly cares, for BMD/BDL probably the same. They just want a dumb extract for files, they don't need to do hash checks. Also yeah, AFAIK the RARC hash is used by BMD/BDL too. Let me know if you have any more questions.
